spec/unit/serializer_spec.rb

```diff
@@ -0,0 +1,214 @@
+require "spec_helper"
+
+RSpec.describe IknowParams::Serializer do
+  shared_examples "a serializer" do
+    it "can load its type from a string" do
+      valid_values.zip(parsed_values).each do |valid, parsed|
+        expect(described_class.load(valid)).to eq parsed
+      end
+    end
+
+    it "raises an error when asked to load a bad value" do
+      invalid_values.each do |invalid|
+        expect { described_class.load(invalid) }.to(raise_error(Exception), "value was accepted: #{invalid}")
+      end
+    end
+
+    it "dumps to a string" do
+      parsed_values.each do |parsed|
+        expect(described_class.singleton.dump(parsed)).to eq parsed.public_send(string_serializer)
+      end
+    end
+
+    it "dumps to a string or JSON value when asked for JSON" do
+      parsed_values.each do |parsed|
+        if described_class.json_value?
+          expect(described_class.singleton.dump(parsed, json: true)).to eq parsed
+        else
+          expect(described_class.singleton.dump(parsed, json: true)).to eq parsed.public_send(string_serializer)
+        end
+      end
+    end
+
+    it "matches the type of arguments" do
+      expect(described_class.singleton.matches_type?(parsed_values.first)).to eq true
+
+      invalid_values.each do |invalid|
+        expect(described_class.singleton.matches_type?(invalid)).to(eq(false), "value was accepted: #{invalid}")
+      end
+    end
+  end
+
+  shared_examples "a JSON-value serializer" do
+    it "represents a JSON value" do
+      expect(described_class.json_value?).to eq true
+    end
+  end
+
+  shared_examples "a non-JSON value serializer" do
+    it "represents a non-JSON value" do
+      expect(described_class.json_value?).to eq false
+    end
+  end
+
+  # What we expect dump to use to serialize the entity to a string
+  let(:string_serializer) { :to_s }
+
+  describe IknowParams::Serializer::Integer do
+    let(:valid_values)  { ["33", 33, "-1", -2] }
+    let(:parsed_values) { [33,   33,  -1,  -2] }
+
+    let(:invalid_values) { ["fish", "4.5", 4.5, Object.new] }
+
+    it_behaves_like "a serializer"
+    it_behaves_like "a JSON-value serializer"
+  end
+
+  describe IknowParams::Serializer::Float do
+    let(:valid_values)  {  ["33.3", 33.3, "0.0", 0.0, 0] }
+    let(:parsed_values) {  [33.3,   33.3,  0.0,  0.0, 0.0] }
+
+    let(:invalid_values) { ["fish", {}, Object.new] }
+
+    it_behaves_like "a serializer"
+    it_behaves_like "a JSON-value serializer"
+  end
+
+  describe IknowParams::Serializer::Boolean do
+    let(:valid_values)  {  [true, 'yes', 'on', 'ON', 'true', '1', 1, 1.0, false, 'no', 'off', 'OFF', 'false', '0', 0, 0.0] }
```

Comment:
It's pretty hard to match up the pairs in two arrays like here. Would it make sense to specify valid/parsed pairs in a hash?